File: Risiko/Assets/Scripts/businesslogic/ClientManager.cs
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using UnityEngine;

namespace businesslogic
{
    public class ClientManager
    {
        private static ClientManager _instance;
        private static readonly object Lock = new object();
        private bool _connected = false;
        private ClientManager() // Private constructor to allow instantiation using singleton only
        {
        }
        public static ClientManager Instance // Implementing singleton pattern
        {
            get
            {
                // Using lock to manage concurrency
                lock (Lock)
                {
                    if (_instance == null)
                    {
                        _instance = new ClientManager();
                    }
                    return _instance;
                }
            }
        }
        private static readonly RequestHandler RequestHandler = new RequestHandler();
        //private string _server = "ws://150.217.51.105:12345"; // 105  
        //private string _server = "ws://93.57.245.63:12345"; // Elia
        //private string _server = "ws://101.58.64.113:12345"; // Filo
        //private string _server = "ws://128.116.252.173:12345"; // Franci
        private string _server = "ws://localhost:12345";
        private ClientWebSocket _webSocket = null;
        private CancellationToken _cancellationToken;
        private bool _isConnectedToLobby = false;
        private List<string> onlineServers = new List<string>();
        public List<string> GetOnlineServersList()
        {
            return onlineServers;
        }
        
        private static readonly List<string> ServersToCheck = new List<string>
        {
            "ws://150.217.51.105:12345", // 105
            "ws://93.57.245.63:12345", // Elia
            "ws://101.58.64.113:12345", // Filo
            "ws://128.116.252.173:12345", // Franci
            "ws://localhost:12345"
        };

        public async Task<List<string>> FetchOnlineServers()
        {
            onlineServers = new List<string>();
            foreach (var server in ServersToCheck)
            {
                if (await IsServerOnline(server))
                {
                    string cleanServer = server.Replace("ws://", "").Replace(":12345", "");
                    onlineServers.Add(cleanServer);
                }
            }
            
            return onlineServers;
        }

        private async Task<bool> IsServerOnline(string serverUri)
        {
            using ClientWebSocket webSocket = new ClientWebSocket();
            var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(4));
            try
            {
                await webSocket.ConnectAsync(new Uri(serverUri), cancellationTokenSource.Token);
                if (webSocket.State == WebSocketState.Open)
                {
                    await webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None);
                    return true;
                }
            }
            catch (Exception ex)
            {
                Debug.Log($"Failed to connect to {serverUri}: {ex.Message}");
            }

            return false;
        }
        public void SetActiveServer(string server)
        {
            _server = "ws://" + server + ":12345";
        }

        public string GetActiveServer()
        {
            return _server;
        }

        public bool IsConnectedToLobby()
        {
            return _isConnectedToLobby;
        }

        public void setIsConnectedToLobby(bool isConnectedToLobby)
        {
            _isConnectedToLobby = isConnectedToLobby;
        }
        public bool IsConnected()
        {
            return _connected;
        }

        public void SetConnected(bool value)
        {
            _connected = value;
        }

        public async void ResetConnection()
        {
            //await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None);
            _connected = false;
            _webSocket = null;
        }

        public async Task StartClient()
        {
            if (_webSocket == null)
            {
                var cancellationTokenSourceFirstConnection = new CancellationTokenSource(TimeSpan.FromSeconds(6));
                var cancellationTokenSource = new CancellationTokenSource();
                var uri = new Uri(_server);
            
                _webSocket = new ClientWebSocket();
            
                try
                {
                    //await _webSocket.ConnectAsync(uri, cancellationTokenSource.Token);
                    _webSocket.Options.KeepAliveInterval = TimeSpan.FromMinutes(2.5); // invia un ping ogni 2.5 minuti
                    await _webSocket.ConnectAsync(uri, cancellationTokenSourceFirstConnection.Token);
            
                    if (_webSocket.State == WebSocketState.Open)
                    {
                        Debug.Log("Connected");
                        SetConnected(true);
                    
                        var handlerTask = RequestHandler.HandleRequests(cancellationTokenSource.Token);
                        var receiveTask = ReceiveMessage(_webSocket, cancellationTokenSource.Token);
                        await Task.WhenAll(handlerTask, receiveTask);
                    }
                    else
                    {
                        Debug.Log("[ClientManager] WebSocket connection could not be established.");
                        _webSocket = null;
                        SetConnected(false);
                    }
                }
                catch (OperationCanceledException)
                {
                    Debug.Log("WebSocket connection was canceled due to timeout.");
                    _webSocket = null;
                    SetConnected(false);
                }
                catch (WebSocketException ex)
                {
                    Debug.Log($"WebSocketException: {ex.Message}");
                    _webSocket = null;
                    SetConnected(false);
                }
                catch (Exception ex)
                {
                    Debug.Log($"Exception: {ex.Message}");
                    _webSocket = null;
                    SetConnected(false);
                }
            }
        }
    
        private static async Task SendMessage(ClientWebSocket webSocket, CancellationToken cancellationToken, string message)
        {
            var buffer = Encoding.UTF8.GetBytes(message);
            await webSocket.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, cancellationToken);
        }
    
        private static async Task ReceiveMessage(ClientWebSocket webSocket, CancellationToken cancellationToken)
        {
            var buffer = new byte[1024 * 24];
            while (!cancellationToken.IsCancellationRequested)
            {
                var result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);
                var response = Encoding.UTF8.GetString(buffer, 0, result.Count);
                RequestHandler.AddRequest(webSocket.Options.ClientCertificates.ToString(), response);
            }
            Debug.Log("Uscito dal loop delle richieste");
        }

        public async void RequestAllGames()
        {
            await SendMessage(_webSocket, _cancellationToken, "SELECT_ALL_GAMES");
        }
        public async void CreateLobbyAsHost()
        {
            await SendMessage(_webSocket, _cancellationToken, "HOST_GAME:"); // Telling the server that I will be the host
        }

        public async void KillLobby()
        {
            setIsConnectedToLobby(false);
            await SendMessage(_webSocket, _cancellationToken, "GAME_KILLED_BY_HOST: " + Player.Instance.PlayerId);
        }

        public async void LeaveLobby()
        {
            setIsConnectedToLobby(false);
            await SendMessage(_webSocket, _cancellationToken, "PLAYER_HAS_LEFT_THE_LOBBY: " + Player.Instance.PlayerId);
        }
    
        public async void LeaveGame()
        {
            setIsConnectedToLobby(false);
            await SendMessage(_webSocket, _cancellationToken, "PLAYER_HAS_LEFT_THE_GAME: " + Player.Instance.PlayerId);
        }
        public async void JoinLobbyAsClient(string lobbyID)
        {
            await SendMessage(_webSocket, _cancellationToken, "JOIN_GAME: " + lobbyID);
        }
    
        public async void RequestNameUpdatePlayerList()
        {
            await SendMessage(_webSocket, _cancellationToken, "REQUEST_NAME_UPDATE_PLAYER_LIST: ");
        }

        public async void SendName()
        {
            await SendMessage(_webSocket, _cancellationToken, "UPDATE_NAME: " + Player.Instance.PlayerId + "-"+  Player.Instance.Name);
        }
    
        public async void StartHostGame()
        {
            await SendMessage(_webSocket, _cancellationToken, "GAME_STARTED_BY_HOST: ");
        }
    
        public async void SendChosenArmyColor()
        {
            await SendMessage(_webSocket, _cancellationToken, "CHOSEN_ARMY_COLOR: " + Player.Instance.PlayerId + "-" + Player.Instance.ArmyColor);
        }

        public async void UpdateTerritoriesState()
        {
            await SendMessage(_webSocket, _cancellationToken, "UPDATE_TERRITORIES_STATE: " + Player.Instance.PlayerId + ", " + JsonConvert.SerializeObject(Player.Instance.Territories));
        }

        public async void AttackEnemyTerritory(Territory myTerritory, Territory enemyTerritory, int myNumArmy)
        {
            int enemyArmyNum;
            if (enemyTerritory.num_tanks >= 3)
            {
                enemyArmyNum = 3;
            }
            else
            {
                enemyArmyNum = enemyTerritory.num_tanks;
            }
        
            GameManager.Instance.setEnemyArmyNum(enemyArmyNum);
            GameManager.Instance.setMyArmyNum(myNumArmy);
            GameManager.Instance.setImUnderAttack(false);
            GameManager.Instance.setMyTerritory(myTerritory);
            GameManager.Instance.setEnemyTerritoy(enemyTerritory);
        
            await SendMessage(_webSocket, _cancellationToken, "ATTACK_TERRITORY: " + 
                                                              Player.Instance.PlayerId + "-" + enemyTerritory.player_id + ", " 
                                                              + myTerritory.id + "-" + enemyTerritory.id + ", " + 
                                                              myNumArmy.ToString() + "-" + enemyArmyNum.ToString());
            //GameManager.Instance.setImAttacking(true); SPOSTATO IN requestHandler
        }
    
        public async void RequestTerritoryInfo(string Terr_id)
        {
            await SendMessage(_webSocket, _cancellationToken, "REQUEST_TERRITORY_INFO: " + Player.Instance.PlayerId + "-" + Terr_id);
    
        }

        public async void RequestAddBot()
        {
            await SendMessage(_webSocket, _cancellationToken, "ADD_BOT");
        }

        public async void RequestRemoveBot()
        {
            await SendMessage(_webSocket, _cancellationToken, "REMOVE_BOT");
        }

    }
}

File: server/bot/Client_manager.py
import websockets
import json
import asyncio

from Player import Player
from Game_manager import GameManager
from RequestHandler import RequestHandler
from Territory import Territory


class ClientManager:

    def __init__(self):
        self._connected = False
        self.player = None
        self._websocket = None
        self.game_manager = GameManager()

    async def start_client(self, ip, port):
        print("Try to connect...")
        async with websockets.connect(f'ws://{ip}:{port}') as websocket:
            try:
                print("debug")
                self._connected = True
                self.player = Player(websocket)
                self._websocket = websocket
                request_handler = RequestHandler(self.player, self.game_manager)
                handler_task = asyncio.create_task(request_handler.handler(websocket))
                print("Connected")
                await asyncio.gather(handler_task)
            except Exception as e:
                await websocket.close()


    def is_connected(self):
        return self._connected

    def set_connected(self, value):
        self._connected = value

    async def reset_connection(self):
        if self._websocket is not None:
            await self._websocket
            self._connected = False
            self._websocket = None

    async def send_message(self, message):
        if self._websocket is not None:
            await self._websocket.send(message)

    async def receive_message(self, websocket):
        async for message in websocket.messages:
            print(f"Received: {message}")
            # Process the message using a RequestHandler (not implemented here)
            # RequestHandler.add_request(websocket.remote_address, message)

    async def create_lobby_as_host(self):
        await self.send_message("HOST_GAME:")

    async def kill_lobby(self):
        await self.send_message(f"LOBBY_KILLED_BY_HOST: {self.player.player_id}")

    async def leave_lobby(self):
        await self.send_message(f"PLAYER_HAS_LEFT_THE_LOBBY: {self.player.player_id}")

    async def join_lobby_as_client(self, lobby_id):
        await self.send_message(f"JOIN_GAME: {lobby_id}")

    async def request_name_update_player_list(self):
        await self.send_message("REQUEST_NAME_UPDATE_PLAYER_LIST: ")

    async def send_name(self):
        print(f'Sending name: {self.player.name}')
        await self.send_message(f"UPDATE_NAME: {self.player.player_id}-{self.player.name}")

    async def start_host_game(self):
        await self.send_message("GAME_STARTED_BY_HOST: ")

    async def send_chosen_army_color(self):
        await self.send_message(f"CHOSEN_ARMY_COLOR: {self.player.player_id}-{self.player.army_color}")

    async def update_territories_state(self):
        territories_dict = [terr.to_dict() for terr in self.player.territories]
        await self.send_message(f"UPDATE_TERRITORIES_STATE: {self.player.player_id}, {json.dumps(territories_dict)}")

    async def attack_enemy_territory(self, my_territory, enemy_territory, my_num_army):
        enemy_num_army = min(3, enemy_territory.num_tanks)
        await self.send_message(f"ATTACK_TERRITORY: {self.player.player_id}-{enemy_territory.player_id}, "
                                f"{my_territory.id}-{enemy_territory.id}, {my_num_army}-{enemy_num_army}")
        GameManager.get_instance().set_im_attacking(True)

    async def request_territory_info(self, terr_id):
        await self.send_message(f"REQUEST_TERRITORY_INFO: {self.player.player_id}-{terr_id}")

    async def request_shortest_path(self, from_terr_node, to_terr_node):
        await self.send_message(f"REQUEST_SHORTEST_PATH: {self.player.player_id}-{from_terr_node}-{to_terr_node}")
